src/types/bone.d.ts

```diff
@@ -0,0 +1,410 @@
+   */
+  static primaryColumn: string;
+
+  /**
+   * The attribute definitions of the model.
+   */
+  static attributes: { [key: string]: AbstractDataType<DataType> | AttributeMeta };
+
+  /**
+   * The schema info of current model.
+   */
+  static columns: Array<AttributeMeta>;
+
+  /**
+   * If the table consists of multiple partition tables then a sharding key is needed to perform actual query. The sharding key can be specified through overridding this property, which will then be used to check the query before it hits database.
+   */
+  static shardingKey: string;
+
+  /**
+   * If the table name is just an alias and the schema info can only be fetched by one of its partition table names, physic tables should be specified.
+   */
+  static physicTables: string[];
+
+  isNewRecord: boolean;
+
+  /**
+   * Override attribute metadata
+   * @example
+   * Bone.attribute('foo', { type: JSON })
+   */
+  static attribute(name: string, meta: AttributeMeta): void;
+
+  /**
+   * Rename attribute
+   * @example
+   * Bone.renameAttribute('foo', 'bar')
+   */
+  static renameAttribute(originalName: string, newName: string): void;
+
+  static alias(name: string): string;
+  static alias(data: Record<string, Literal>): Record<string, Literal>;
+  static unalias(name: string): string;
+
+  static hasOne(name: string, opts?: AssociateOptions): void;
+  static hasMany(name: string, opts?: AssociateOptions): void;
+  static belongsTo(name: string, opts?: AssociateOptions): void;
+
+  /**
+   * INSERT rows
+   * @example
+   * Bone.create({ foo: 1, bar: 'baz' })
+   */
+  static create<T extends typeof AbstractBone>(this: T, values: Values<T>, options?: QueryOptions): Promise<InstanceType<T>>;
+
+  /**
+   * INSERT or UPDATE rows
+   * @example
+   * Bone.upsert(values, { hooks: false })
+   * @param values values
+   * @param opt query options
+   */
+  static upsert<T extends typeof AbstractBone>(this: T, values: Object, options?: QueryOptions): Spell<T, number>;
+
+  /**
+   * Batch INSERT
+   */
+  static bulkCreate<T extends typeof AbstractBone>(this: T, records: Array<Record<string, Literal>>, options?: QueryOptions): Promise<Array<InstanceType<T>>>;
+
+  /**
+   * SELECT all rows. In production, when the table is at large, it is not recommended to access records in this way. To iterate over all records, {@link Bone.batch} shall be considered as the better alternative. For tables with soft delete enabled, which means they've got `deletedAt` attribute, use {@link Bone.unscoped} to discard the default scope.
+   */
+  static all: Spell<typeof AbstractBone, Collection<AbstractBone>>;
+
+  /**
+   * SELECT rows OFFSET index LIMIT 1
+   * @example
+   * Bone.get(8)
+   * Bone.find({ foo: { $gt: 1 } }).get(42)
+   */
+  static get<T extends typeof AbstractBone>(this: T, index: number): Spell<T, InstanceType<T> | null>;
+
+  /**
+   * SELECT rows ORDER BY id ASC LIMIT 1
+   */
+  static first: Spell<typeof AbstractBone, AbstractBone | null>;
```

Comment:
Didn't get a proper way to fix this type issue(same as `await Bone.all/last`、 `Bone.unscoped` and `SequelizeBone.Model`  )
```ts
class Post extends Bone {
   @Column()
   title: string;
}

const p = await Post.first;
// error here, title not a property of post
p.title;
```

Reply:
`await Post.find().first` can work temporarily because of `Spell.first/last`

Reply:
yes, static getter in TypeScript seems can't have polymorphic this. https://github.com/microsoft/TypeScript/issues/50301